pkg/providers/cloudstack/config/template-cp.yaml

```
@@ -0,0 +1,305 @@
+      path: "/etc/containerd/config_append.toml"
+{{- end }}
+{{- if .awsIamAuth}}
+    - content: |
+        # clusters refers to the remote service.
+        clusters:
+          - name: aws-iam-authenticator
+            cluster:
+              certificate-authority: /var/aws-iam-authenticator/cert.pem
+              server: https://localhost:21362/authenticate
+        # users refers to the API Server's webhook configuration
+        # (we don't need to authenticate the API server).
+        users:
+          - name: apiserver
+        # kubeconfig files require a context. Provide one for the API Server.
+        current-context: webhook
+        contexts:
+        - name: webhook
+          context:
+            cluster: aws-iam-authenticator
+            user: apiserver
+      permissions: "0640"
+      owner: root:root
+      path: /var/lib/kubeadm/aws-iam-authenticator/kubeconfig.yaml
+    - contentFrom:
+        secret:
+          name: aws-iam-authenticator-ca
+          key: cert.pem
+      permissions: "0640"
+      owner: root:root
+      path: /var/lib/kubeadm/aws-iam-authenticator/pki/cert.pem
+    - contentFrom:
+        secret:
+          name: aws-iam-authenticator-ca
+          key: key.pem
+      permissions: "0640"
+      owner: root:root
+      path: /var/lib/kubeadm/aws-iam-authenticator/pki/key.pem
+{{- end}}
+    initConfiguration:
+      nodeRegistration:
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs:
+          provider-id: cloudstack:///'{{`{{ ds.meta_data.instance_id }}`}}'
+          read-only-port: "0"
+          anonymous-auth: "false"
+        name: '{{`{{ ds.meta_data.local_hostname }}`}}'
+    joinConfiguration:
+      nodeRegistration:
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs:
+          provider-id: cloudstack:///'{{`{{ ds.meta_data.instance_id }}`}}'
+          read-only-port: "0"
+          anonymous-auth: "false"
+        name: '{{`{{ ds.meta_data.local_hostname }}`}}'
+    preKubeadmCommands:
+    - swapoff -a
+{{- if.registryMirrorConfiguration }}
+    - cat /etc/containerd/config_append.toml >> /etc/containerd/config.toml
+{{- end }}
+{{- if or .proxyConfig .registryMirrorConfiguration }}
+    - sudo systemctl daemon-reload
+    - sudo systemctl restart containerd
+{{- end }}
+    - hostname "{{`{{ ds.meta_data.local_hostname }}`}}"
+    - echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
+    - echo "127.0.0.1   localhost" >>/etc/hosts
+    - echo "127.0.0.1   {{`{{ ds.meta_data.local_hostname }}`}}" >>/etc/hosts
+    - echo "{{`{{ ds.meta_data.local_hostname }}`}}" >/etc/hostname
+    useExperimentalRetryJoin: true
+    users:
+    - name: {{.controlPlaneSshUsername}}
+      sshAuthorizedKeys:
+      - '{{.cloudstackControlPlaneSshAuthorizedKey}}'
+      sudo: ALL=(ALL) NOPASSWD:ALL
+    format: {{.format}}
+  replicas: {{.controlPlaneReplicas}}
+  version: {{.kubernetesVersion}}
+---
+{{- if .externalEtcd }}
```

```suggestion
{{- if .externalEtcd }}
---
```
Similar to the comment below, but do we need these trailing delimiters if our templater will add them anyways?

---

I'm not sure if I follow. I was under the impression the templater just evaluates the logic and fills in variable as specified. Are you concerned about having duplicate delimiters, such as in the case where .exteralEtcd is undefined?

---

Yup mainly about duplicate delimiters, which might be a problem in vsphere too. Might be better to catch in the review that we have here, which was why I listed it as a comment

---

Refactored :)